pkg/distributor/distributor.go

```diff
@@ -1223,12 +1226,27 @@ func (d *Distributor) prePushValidationMiddleware(next PushFunc) PushFunc {
 			skipLabelCountValidation := d.cfg.SkipLabelCountValidation || req.GetSkipLabelCountValidation()
 
 			// Note that validateSeries may drop some data in ts.
+			rawSamples := len(ts.Samples)
+			rawHistograms := len(ts.Histograms)
 			shouldRemove, validationErr := d.validateSeries(now, &req.Timeseries[tsIdx], userID, group, skipLabelValidation, skipLabelCountValidation, minExemplarTS, maxExemplarTS)
 
 			if countDroppedNativeHistograms {
 				droppedNativeHistograms += len(ts.Histograms)
 			}
 
+			dedupedSamplesAndHistograms := (rawSamples - len(ts.Samples)) + (rawHistograms - len(ts.Histograms))
+			if dedupedSamplesAndHistograms > 0 {
+				for _, l := range ts.Labels {
+					if l.Name != labels.MetricName {
+						continue
+					}
+					if _, exists := dedupedPerMetric[l.Value]; exists || len(dedupedPerMetric) < maxTrace {
+						dedupedPerMetric[l.Value] += dedupedSamplesAndHistograms
```

Comment:
The strings here point into a large gRPC buffer, so care needs to be taken not to retain them.
I don't think they get retained here, but some comment to this effect is probably a good idea.

Reply:
Also, `SpanLogger.DebugLog` logs the arguments if in debug mode. Isn't there a risk that the logger retains the unsafe strings until flushed?